Dspace 7 shibboleth (REST)

atarix83 opened the pull request:
This PR provide changes to make the shibboleth authentication work as discussed in [DS-4396](https://jira.lyrasis.org/browse/DS-4396)

A summary of what is done with this PR:

- add a new ShibbolethRestController 
- add a new ShibbolethAuthenticationFilter
- add the possibility to check a JWT token saved in a http cookie sent from client

As discussed in the issue [DS-4396](https://jira.lyrasis.org/browse/DS-4396) this implementation doesn't handle the `Access-Control-Allow-Origin` header in the response to avoid CORS problem.

To solve it, we use in the rest VH the following apache configuration : 
```
# enable CORS headers for dspace7-demo.atmire.com and localhost
SetEnvIf Origin "^http(s)?://(www\.)?(dspace7\-demo\.atmire\.com|localhost|127\.0\.0\.1)$" AccessControlAllowOrigin=$0
  Header set Access-Control-Allow-Origin %{AccessControlAllowOrigin}e env=AccessControlAllowOrigin
  Header set Access-Control-Allow-Credentials true env=AccessControlAllowOrigin
  Header set Access-Control-Expose-Headers: "Authorization, expires, Location, Content-Disposition, WWW-Authenticate, Set-Cookie, X-Requested-With"
  Header merge Vary Origin
```

this PR has been already merged in the dspace-cris 7 codebase so it is possible to see the behavior here: [https://dspacecris7.4science.cloud/home](https://dspacecris7.4science.cloud/home)

atarix83 committed:
Added a cookie with JWT to response

atarix83 committed:
Handle redirect to client after shibboleth authentication succeded

atarix83 committed:
Added login filter to /api/authn/shibboleth

atarix83 committed:
Added shibboleth authentication filter

atarix83 committed:
Chack if authentication cookie is present in Authentication Data

atarix83 committed:
use X-Requested-With header if Referer is empty

atarix83 committed:
Merge remote-tracking branch 'origin/master' into dspace-7-shibboleth

atarix83 committed:
Fixed checkstyle violations

atarix83 committed:
Retrieve token from cookie only when checking an authenticated eperson

atarix83 committed:
Revert "Retrieve token from cookie only when checking an authenticated eperson"

This reverts commit 10bbf8079b3f54be7b60d5ed973605dafb6d73b8.

atarix83 committed:
Add authorization cookie only in ShibbolethAuthenticationFilter

atarix83 committed:
Fixed AuthenticationRestControllerIT

atarix83 committed:
Fixed checkstyle violations

atarix83 committed:
Invalidate authorization cookie on logout

atarix83 committed:
Revert "add context commit after invalidating token"

This reverts commit 64bc25f7c9f9e03f878809982235022bbe3af339.

atarix83 committed:
add context commit after invalidating token

atarix83 committed:
use authorization cookie only to check

atarix83 committed:
fix use authorization cookie only to check

atarix83 committed:
Invalidate authentication cookie once used

atarix83 committed:
Add checck to user and password blank

atarix83 committed:
fix check user and password blank

atarix83 committed:
Revert check

atarix83 committed:
Check session salt in shib auth

atarix83 committed:
Revert "Check session salt in shib auth"

This reverts commit 462f20a3389b6e9641a9af7f6992f3266bce6b45.

atarix83 committed:
add attemptAuthentication method to ShibbolethAuthenticationFilter

atarix83 committed:
fix attemptAuthentication method on ShibbolethAuthenticationFilter

atarix83 committed:
Added WWW-Authenticate header to authn/status response

atarix83 committed:
Revert "use authorization cookie only to check"

atarix83 committed:
Merge remote-tracking branch 'origin/master' into dspace-7-shibboleth

atarix83 committed:
replace Boolean type with primitive type

atarix83 committed:
use invalidateAuthenticationCookie method

atarix83 committed:
Merge remote-tracking branch 'origin/master' into dspace-7-shibboleth

benbosman reviewed (commented):
Thanks @atarix83 for this PR

This looks good, I only have some minor feedback inline

Can you also add ITs for:
* SHIB_AND_PASS (verifying both work)
* Verify using PASS_ONLY that password authentication works
* Verify using PASS_ONLY that shibboleth authentication doesn't work
* Verify using SHIB_ONLY that password authentication doesn't work
* Verify using SHIB_ONLY that password authentication works

atarix83 committed:
Added setSecure option to authetication cookie

atarix83 committed:
Added inline comment

atarix83 committed:
Added more integration test for shibboleth authentication

atarix83 committed:
Added integration test fo ShibbolethRestController

atarix83 committed:
Fixed the default value of authentication-shibboleth.lazysession.loginurl param

paulo-graca commented on dspace-api/src/main/java/org/dspace/authenticate/ShibAuthentication.java:
```diff
@@ -509,19 +510,17 @@ public String loginPageURL(Context context, HttpServletRequest request, HttpServ
+                redirectUrl = request.getHeader("X-Requested-With");
             }
 
+            String returnURL = ConfigurationManager.getProperty("dspace.baseUrl") + "/api/authn/shibboleth"
```
Please note that this property it's about to be changed to: `dspace.server.url`

tdonohue replied:
Yes, this config now needs to be updated to `dspace.server.url`, as #2657 was merged

paulo-graca commented on dspace/config/modules/authentication-shibboleth.cfg:
```diff
@@ -38,7 +38,7 @@
 
 # The url to start a shibboleth session (only for lazy sessions)
-authentication-shibboleth.lazysession.loginurl = /Shibboleth.sso/Login
+authentication-shibboleth.lazysession.loginurl = ${dspace.baseUrl}/Shibboleth.sso/Login
```
Please note that this property it's about to be changed to: dspace.server.url

benbosman replied:
We have to be careful which property is used here. It should not redirect to /server/Shibboleth.sso/Login (which it currently does in my installation)

I'd recommend to revisit this once https://github.com/DSpace/DSpace/pull/2657 is merged because there's a risk the URL is invalid

tdonohue replied:
@atarix83 : I'm unclear why this change is needed.  If we leave this setting as it was (`/Shibboleth.sso/Login`), that should cause the redirect to be to the servlet container root path.  That means if you were running DSpace REST API at http://localhost:8080/server/ then the path "/Shibbboleth.sso/Login" should cause the redirect to be to http://localhost:8080/Shibboleth.sso/Login.  

I believe that's how this setting worked in DSpace 6 and it also is how relative paths are described to work in HttpServletResponse.sendRedirect(): https://tomcat.apache.org/tomcat-8.5-doc/servletapi/javax/servlet/http/HttpServletResponse.html?is-external=true#sendRedirect(java.lang.String)  (This is the method that ultimately gets called with this relative URL)

So, I guess I'm asking why we need to turn this configuration into an absolute/full URL, as it worked previously as a relative URL?

benbosman replied:
In DSpace 7, the difference is that the hostname of REST and Angular may differ
When logging in from the Angular UI, Angular should link to https://dspace7-rest/Shibboleth.sso/Login (not https://dspace7-rest/server/Shibboleth.sso/Login nor https://dspace7-angular/Shibboleth.sso/Login)

At this point in time, this would requiring taking ${dspace.server.url}, stripping the context path off, and adding /Shibboleth.sso/Login to it. That makes it hard to configure

paulo-graca reviewed (commented):
Thank you @atarix83! I'm still doing some tests and this PR seems to be in a good way. Please mind that you are using dspace configs that are about to be changed.

benbosman commented:
I've re-reviewed this branch, and the only issue I noticed now was with the shibboleth lazy session login URL which redirected me to /server/Shibboleth.sso/Login now.
If this can be resolved, I can test once more with the UI and REST on a separate URL using the CORS headers
I have one more question: Is it possible to implement the CORS headers in REST for the login URL endpoint (instead of depending on an Apache config)? I didn't verify how the CORS wildcard headers are currently set in REST.

atarix83 committed:
Fixed checkstyle violations

atarix83 commented:
@benbosman 

> I've re-reviewed this branch, and the only issue I noticed now was with the shibboleth lazy session login URL which redirected me to /server/Shibboleth.sso/Login now.

You are right I've changed the default value [here](https://github.com/DSpace/DSpace/pull/2651/files#diff-7af0c4da25387f79141b778592fa58fbR41) but I realized that `dspace.baseUrl` can't be use because it contain `/server`, I'll fix it in an other way. Anyway if you want to test you can simply overwrite the default value with the correct url in your local.cfg

> If this can be resolved, I can test once more with the UI and REST on a separate URL using the CORS headers


I put here the right apache configuration that allows to login with localhost too : 
```
# enable CORS headers for dspace7-demo.atmire.com and localhost
SetEnvIf Origin "^http(s)?://(www\.)?(dspace7\-demo\.atmire\.com|localhost\:3000|127\.0\.0\.1\:3000)$" AccessControlAllowOrigin=$0
  Header set Access-Control-Allow-Origin %{AccessControlAllowOrigin}e env=AccessControlAllowOrigin
  Header set Access-Control-Allow-Credentials true env=AccessControlAllowOrigin
  Header set Access-Control-Expose-Headers: "Authorization, expires, Location, Content-Disposition, WWW-Authenticate, Set-Cookie, X-Requested-With"
  Header merge Vary Origin
```
it must contain the 3000 port to avoid CORS errors when you test on `http://localhost:3000` 

> I have one more question: Is it possible to implement the CORS headers in REST for the login URL endpoint (instead of depending on an Apache config)? I didn't verify how the CORS wildcard headers are currently set in REST.

The plan is to have a final solution where REST will handle CORS header, but as decided [here](https://jira.lyrasis.org/browse/DS-4396?focusedCommentId=63627&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-63627) this first implementation use the apache configuration as temporary workaround

paulo-graca commented:
I've managed to get Shibboleth authentication (version 3) working with our IdP.
One thing I would like to share with others since I struggled a bit to get it working (and thanks @atarix83 for all the help) is related my apache configuration. I'm using mod proxy to use the same machine with DSpace angular and DSpace Server. One thing I wasn't doing was enable shibboleth sitewide. So, after authentication, with the redirects, I lost the access to the attributes. I've something like this. I hope it help others.

```
     ProxyPass /server ajp://localhost:8009/server
     ProxyPass /Shibboleth.sso !

    <Location />
        Order deny,allow
        Allow from all
        AuthType shibboleth
        ShibUseHeaders On
        Require shibboleth
    </Location>
```

paulo-graca commented:
One feature I would like to see in this PR it's the ability to logout. Currently the logout process is only a local session termination. After it, if I again click on shibboleth authentication, the old session is reused. This is a big issue for shared computers. I will need a dig a bit and will try to provide more feedback.

mwoodiupui commented:
Logout:  please don't make the mistake that we made here with CAS.  Here everything has a single logout button that invalidates the CAS session.  Thus if I logout of anything I'm logged out of everything, which is almost always *not* what I wanted.  This has trained me to never logout from CASified applications.  And *that* has taught me that SSOified applications may need two logout buttons:  invalidate this application session, and (optionally) invalidate my SSO session.

tdonohue commented on dspace-api/src/main/java/org/dspace/authenticate/ShibAuthentication.java:
```diff
@@ -33,6 +33,7 @@
 import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.MetadataFieldService;
 import org.dspace.content.service.MetadataSchemaService;
+import org.dspace.core.ConfigurationManager;
```
Just a note here. You shouldn't need to import the ConfigurationManager, as this class already has access to the `ConfigurationService`.  The ConfigurationManager is deprecated in favor of the ConfigurationService.  So, you should remove this import and just use `configurationService` instead.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/ShibbolethRestController.java:
```diff
@@ -0,0 +1,58 @@
+    public void shibboleth(HttpServletResponse response,
+            @RequestParam(name = "redirectUrl", required = false) String redirectUrl) throws IOException {
+        if (redirectUrl == null) {
+            redirectUrl = ConfigurationManager.getProperty("dspace.url");
```
This configuration needs updating as well to be `dspace.ui.url`.  Also, you should use ConfigurationService (which can be `@Autowired` like other services)

paulo-graca reviewed (commented):
> Logout: please don't make the mistake that we made here with CAS. Here everything has a single logout button that invalidates the CAS session. Thus if I logout of anything I'm logged out of everything, which is almost always _not_ what I wanted. This has trained me to never logout from CASified applications. And _that_ has taught me that SSOified applications may need two logout buttons: invalidate this application session, and (optionally) invalidate my SSO session.

With our services we have a different reality. We use a national identification service, that also integrates with a government id service, and upon logout, all sessions should be terminated (across all service providers). For future work, I think this can and should be configured. Whether, or not, use a SSO logout.

paulo-graca commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/security/CustomLogoutHandler.java:
```diff
@@ -45,7 +45,7 @@ public void logout(HttpServletRequest httpServletRequest, HttpServletResponse ht
         try {
             Context context = ContextUtil.obtainContext(httpServletRequest);
-            restAuthenticationService.invalidateAuthenticationData(httpServletRequest, context);
+            restAuthenticationService.invalidateAuthenticationData(httpServletRequest, httpServletResponse, context);
```
A verification can be used to identify if the current session is a shibboleth session and then call the configured logout parameter

paulo-graca reviewed (commented):
As promised, I will try to summarize what we have done in order to have Logout within shibboleth and DSpace.
Basically, in terms of the Service Provider, we use the shibboleth's notify parameter to configure the IdP logout url:
https://wiki.shibboleth.net/confluence/display/SP3/Notify

`<Notify Channel="front" Location="https://[MY-SERVER]/logout" />`

When the logout event in DSpace is performed, then the Notify is triggered. 
Some exchanges will occur between the SP and the IdP and an indication to IdP to clean the shibboleth authentication will also occur.

In our case we will have something like this: `https://[MY-SERVER]/shibboleth.sso/Logout?notifying=1&index=1`

To invoke the logout. We also clean local cookies too.

benbosman commented on dspace/config/modules/authentication-shibboleth.cfg:
```diff
@@ -38,7 +38,7 @@
 
 # The url to start a shibboleth session (only for lazy sessions)
-authentication-shibboleth.lazysession.loginurl = /Shibboleth.sso/Login
+authentication-shibboleth.lazysession.loginurl = ${dspace.baseUrl}/Shibboleth.sso/Login
```
If it's not specified, it would be a local-only logout.
This will leave at least system administrators with the choice on how to setup their DSpace (based on policies)
This would be a redirect

benbosman commented:
@paulo-graca 

You won't need to expose the Shibboleth header on entire REST as you specified above:

> ```
>      ProxyPass /server ajp://localhost:8009/server
>      ProxyPass /Shibboleth.sso !
> 
>     <Location /server>
>         Order deny,allow
>         Allow from all
>         AuthType shibboleth
>         ShibUseHeaders On
>         Require shibboleth
>     </Location>
> ```

If you want to limit what uses the Shibboleth headers, /server/api/authn/shibboleth and /server/api/authn/login should suffice

atarix83 committed:
Merge remote-tracking branch 'origin/master' into dspace-7-shibboleth

atarix83 committed:
Added getBaseUrl to Utils

atarix83 committed:
Fixed default value of authentication-shibboleth.lazysession.loginurl

atarix83 committed:
Replace ConfigurationManager with ConfigurationService

benbosman reviewed (commented):
Thanks @atarix83 for the changes

I've tested the REST part (without Angular), and it worked correctly

I've reviewed the last changes as well, and I had a few questions for this:
* If you'd be hosting on a different port number (not likely for PROD environments of course), I assume the `getBaseUrl` function won't work correctly. Can you:
  * add the port number in the `getBaseUrl` function
  * test the returned location is correct in AuthenticationRestControllerIT if `authentication-shibboleth.lazysession.loginurl` is modified to contain a port number
  * test the returned location is correct in AuthenticationRestControllerIT if `authentication-shibboleth.lazysession.loginurl` is modified to contain a full URL with hostname included)
* The `getShibURL` function doesn't verify the behavior if the URL is set to `Shibboleth.sso/Login`. Can you either adjust this in the `getShibURL` function or add some comments in authentication-shibboleth.cfg explaining the allowed formats

tdonohue commented on dspace-api/src/main/java/org/dspace/authenticate/ShibAuthentication.java:
```diff
@@ -1258,6 +1246,28 @@ protected String findSingleAttribute(HttpServletRequest request, String name) {
+        if (shibURL == null || shibURL.length() == 0) {
+            shibURL = "/Shibboleth.sso/Login";
+        }
+        shibURL = shibURL.trim();
```
A small comment.  You don't need to do all this string parsing of `shibURL`, as the ConfigurationService is much "smarter" than the old ConfigurationManager.  The string will already be trimmed.  Additionally, the `if` just before this is unneeded, as you can pass in a default value if it's unspecified.  So, both the `if` and the `trim()` can be removed in favor of just calling

`String shibURL = configurationService.getProperty("authentication-shibboleth.lazysession.loginurl", "/Shibboleth.sso/Login");`

atarix83 replied:
done

tdonohue commented on dspace-api/src/test/java/org/dspace/core/UtilsTest.java:
```diff
@@ -22,6 +22,27 @@
+                     Utils.getBaseUrl("https://dspace.org/"));
+
+        assertEquals("Test keep url", "https://demo.dspace.org",
+                     Utils.getBaseUrl("https://demo.dspace.org"));
```
@atarix83  What would happen if the Base URL includes a port number?  While it's less frequently seen, I have seen sites that run their DSpace on port 8080, for example.  It looks to me like `Utils.getBaseUrl()` would fail if the port is not either 80 (HTTP) or 443 (HTTPS)

atarix83 replied:
added compatibility with url that contains port

atarix83 committed:
Merge remote-tracking branch 'origin/master' into dspace-7-shibboleth

atarix83 committed:
Added port to getBaseUrl

atarix83 committed:
Use configurationService to set default authentication-shibboleth.lazysession.loginurl in getShibURL method

atarix83 committed:
Added methods to test shibboleth url using port

atarix83 committed:
Added comments for authentication-shibboleth.lazysession.loginurl in authentication-shibboleth.cfg

atarix83 commented:
@benbosman 
I've done changes requested

atarix83 commented:
@benbosman Did you tested shibboleth authentication on a server after the upgrade to java 11 ? I've tried to install this PR but it doesn't work, indeed after been logged in with shibboleth on redirect to dspace I get a 403 forbidden. This is not related to this implementation because the same behaviour occurs also on the  [demo rest server](https://dspace7.4science.cloud/server/login.html).
Could you let me know if that works for you?

paulo-graca reviewed (approved):
Thank you @atarix83 for this important contribution! It's working as expected!
I Can see a 
"WWW-Authenticate: 
shibboleth realm="DSpace REST API", 
,password realm="DSpace REST API"

paulo-graca commented:
I've just created: https://jira.lyrasis.org/browse/DS-4464 to address the need of a Central Authentication Service Logout solution.

atarix83 committed:
Merge remote-tracking branch 'origin/master' into dspace-7-shibboleth

atarix83 committed:
Fixed checkstyle violations

benbosman commented:
@atarix83 how is the https://dspace7.4science.cloud currently configured to work with Shibboleth:
* Are you using Apache 2.2, 2.4?
* Are you using mod_http or mod_ajp to proxy to tomcat
* Which config are you using in shib.conf?

I noticed I used to be able to use Shibboleth with the previous codebase: that was based on Apache 2.2, mod_ajp
I now use a new server, containing a more recent version of Java, tomcat, Apache 2.4, mod_ajp, and I am encountering the same issue. It seems like there's a conflict with the setup

@paulo-graca which setup are you using the test the Shibboleth integration?

benbosman reviewed (approved):
This works for me, I've also updated https://wiki.lyrasis.org/display/DSPACE/DSpace+7+Shibboleth+Configuration with my setup and the issue I encountered.

The only problem I still see is the warning I get in Chrome when using separate hostnames for REST and Angular:
> A cookie associated with a cross-site resource at http://dspace7-rest.atmire.com/ was set without the `SameSite` attribute. A future release of Chrome will only deliver cookies with cross-site requests if they are set with `SameSite=None` and `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5088147346030592 and https://www.chromestatus.com/feature/5633521622188032.

This can also be resolved in a follow-up PR

atarix83 commented:
@benbosman 
We are using Apache 2.4, tomcat 9.0.31 and mod_jk to proxy to tomcat.
How did you solve it?

benbosman commented:
@atarix83 I don't use mod_jk, but rather mod_ajp
What solved the problem for me was to remove the attribute attributePrefix="AJP_" in the ApplicationDefaults of shibboleth2.xml
AFAIK, after logging in with Shibboleth, your requests no longer are sent to Tomcat, and I did encounter that problem as well when that attribute was present.

If that doesn't work, can you share your config similar to what Paulo did on the wiki?

tdonohue commented:
Looks good to me, and I see this is already at +2 from Paulo & Ben.  Instructions for setup are drafted at https://wiki.lyrasis.org/display/DSPACE/DSpace+7+Shibboleth+Configuration
